### PR TITLE
Issue/styled switch

### PR DESCRIPTION
**Pull request by xuhaibahmad**
Added styled switch along with a comprehensive launcher activity that lists all the demo's available.

image

image

### CODE REVIEW

File: app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt [hunk 1]
```diff
@@ -21,3 +53,34 @@ class DemoActivity : AppCompatActivity() {
...
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
+        ViewHolder(LayoutInflater.from(parent.context).inflate(R.layout.demo_recycler_view_item, parent, false))
+
+    override fun getItemCount(): Int = items.size
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        holder.bind(items[position])
+        holder.setOnClickListener(clickListener)
+
```
Comment (boyw165) [inline app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt]: MINOR: Remove this empty line.

File: app-demo/src/main/res/layout/activity_demo.xml [hunk 2]
```diff
@@ -3,152 +3,36 @@
     xmlns:android="http://schemas.android.com/apk/res/android"
     xmlns:app="http://schemas.android.com/apk/res-auto"
     android:layout_width="match_parent"
-    android:layout_height="match_parent">
-
-    <android.support.constraint.Guideline
-        android:id="@+id/guide_toolbar"
+    android:layout_height="match_parent"
+    xmlns:tools="http://schemas.android.com/tools">
+
```
Comment (boyw165) [inline app-demo/src/main/res/layout/activity_demo.xml]: MINOR: Remove the two empty lines.

File: app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt [hunk 3]
```diff
@@ -21,3 +53,34 @@ class DemoActivity : AppCompatActivity() {
...
+    }
+
+    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
+
+        private var clickListener: ((view: View, item: DemoViewModel) -> Unit)? = null
+
+        fun bind(viewModel: DemoViewModel) {
+            itemView.findViewById<TextView>(R.id.tvLabel).text = viewModel.name
```
Comment (boyw165) [inline app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt]: VERY MINOR: Just to know we could simplify this by adapting Kotlin Android extension. Could [check out the document here](https://kotlinlang.org/docs/tutorials/android-plugin.html).

File: styled-switch-view/deploy.gradle [hunk 4]
```diff
@@ -0,0 +1,23 @@
+apply plugin: 'com.novoda.bintray-release'
+
+publish {
+    bintrayUser = System.getenv('BINTRAY_USER')
+    bintrayKey = System.getenv('BINTRAY_API_KEY')
+
+    repoName = 'android'
+    uploadName = 'styled-switch-view'
```
Comment (boyw165) [inline styled-switch-view/deploy.gradle]: 🤟Thanks for taking care of this as well. Do you want to set up the Bintray or me?

File: styled-switch-view/src/main/java/co/sodalabs/view/StyledSwitch.kt [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+                val endTime = System.currentTimeMillis()
+                val span = endTime - startTime
+                if (span < 200) {
+                    performClick()
+                } else {
+                    if (x >= mWidth.ushr(1)) {
+                        val switchColor = ValueAnimator.ofInt(if (x > mWidth - padding - thumbRadii) mWidth - padding - thumbRadii else x.toInt(), mWidth - padding - thumbRadii)
+                        switchColor.run {
+                            addUpdateListener { animation ->
+                                val value = animation.animatedValue as Float
```
Comment (boyw165) [inline styled-switch-view/src/main/java/co/sodalabs/view/StyledSwitch.kt]: There is a casting exception here. Error log is like,

```
    java.lang.ClassCastException: java.lang.Integer cannot be cast to java.lang.Float
        at co.sodalabs.view.StyledSwitch$onTouchEvent$$inlined$run$lambda$2.onAnimationUpdate(StyledSwitch.kt:416)
```

File: styled-switch-view/src/main/java/co/sodalabs/view/StyledSwitch.kt [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+                            interpolator = AccelerateDecelerateInterpolator()
+                            duration = 250
+                            start()
+                        }
+                        isOn = true
+                    } else {
+                        val switchColor = ValueAnimator.ofInt(if (x < padding) padding else x.toInt(), padding)
+                        switchColor.run {
+                            addUpdateListener { animation ->
+                                val value = animation.animatedValue as Float
```
Comment (boyw165) [inline styled-switch-view/src/main/java/co/sodalabs/view/StyledSwitch.kt]: Same issue as above.

File: styled-switch-view/src/main/res/values/attrs_styled_switch.xml [hunk 7]
```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+    <declare-styleable
+        name="StyledSwitch">
```
Comment (boyw165) [inline styled-switch-view/src/main/res/values/attrs_styled_switch.xml]: Update the attribute name as the name is updated.

**Comment by boyw165** [comment]
Good job man!

File: app-demo/src/main/AndroidManifest.xml [hunk 8]
```diff
@@ -28,5 +28,15 @@
...
         </activity>
+
+        <!-- RecyclerView Demo -->
+        <activity
+            android:name=".StyledRecyclerViewDemo">
+        </activity>
+
+        <!-- Switch Demo -->
+        <activity
+            android:name=".StyledSwitchDemo">
```
[the remark this answers is not in the document]
Comment (xuhaibahmad) [reply app-demo/src/main/AndroidManifest.xml]: Yup! Good idea.

File: app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt (hunk 1 above)
Comment (boyw165) [reply app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt]: I'd be very happy to have it. Is it possible to integrate the lint quickly?

File: styled-switch-view/src/main/res/values/attrs_styled_switch.xml (hunk 7 above)
Comment (boyw165) [reply styled-switch-view/src/main/res/values/attrs_styled_switch.xml]: Oh, just to name it as `StyledSwitchView`. It's quite minor, up to you. :)

File: app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt (hunk 1 above)
Comment (xuhaibahmad) [reply app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt]: Yup! Ktlint is really easy to setup. Just run the command from here:
https://github.com/shyiko/ktlint#installation and it'll take care of rest.

File: styled-switch-view/src/main/res/values/attrs_styled_switch.xml (hunk 7 above)
Comment (xuhaibahmad) [reply styled-switch-view/src/main/res/values/attrs_styled_switch.xml]: Sure thing!

File: app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt (hunk 1 above)
Comment (boyw165) [reply app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt]: Let's leave the ktlint integrate another ticket.

File: app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt (hunk 3 above)
Comment (boyw165) [reply app-demo/src/main/java/co/sodalabs/demo/DemoActivity.kt]: K, don't worry about it though. 👷‍♂️

File: app-demo/src/main/res/layout/activity_demo_styled_switch.xml [hunk 9]
```diff
@@ -30,6 +33,12 @@
     <co.sodalabs.view.StyledSwitch
         android:layout_width="wrap_content"
         android:layout_height="wrap_content"
+        app:on="false"
```
Comment (boyw165) [inline app-demo/src/main/res/layout/activity_demo_styled_switch.xml]: Could make the label more specific by adding the preview, such as `sw`. As we have more custom attributes, the management of the naming conflict becomes important. To minimize the naming conflict, we could simply add the prefix to the custom attributes. I know the attributes are already scoped under `StyledSwitch`, and it is scoped programmatically. Unfortunately, in XML, it is likely to conflict unless we specify the namespace.

Comment (boyw165) [reply app-demo/src/main/res/layout/activity_demo_styled_switch.xml]: In addition to that, I quite like the `textOn` and `textOff` serving specifically for the label.

File: styled-switch-view/src/main/java/co/sodalabs/view/StyledSwitch.kt (hunk 5 above)
Comment (boyw165) [reply styled-switch-view/src/main/java/co/sodalabs/view/StyledSwitch.kt]: It seems like the `animatedValue` is Int and couldn't be cast to Float with `as` operator.

**Pull request merged**